Resolve operator ambiguities and improve compiler compatibility

pcaddict opened the pull request:
#### Summary

This change enables the use of C++20 compatible toolchains by resolving the following issues:

1. Resolves an ambiguous comparison between `ConcreteAttributePath and ConcreteDataAttributePath`.
	 - The expression `(ConcreteAttributePath) == (ConcreteDataAttributePath)` became ambiguous because the C++20 standard changed the rules for overload resolution for symmetric operators. The compiler finds a "normal" and "reversed" candidate and declares the comparison ambiguous.
2. Replaces compound assignment: In `CHIPCryptoPALPSA.cpp`, it refactors a compound assignment (|=) to its explicit form (= ... |) to prevent potential compiler warnings related to compound assignment to volatile.
3. Update CMake to enable correct C++ standard (Optional/WIP): Modify `enable-gnu-std.cmake` to conditionally select between gnu++17 and gnu++20. This change is needed to fully allow compilation with C++20 because `enable-gnu-std.cmake` unconditionally sets `-std=gnu++17`, potentially overriding the developer-configured C++ standard and results in build errors.
	- **This change will require further discussion. Local testing has begun.**
#### Related issues
Fixes: [Issue 40781](https://github.com/project-chip/connectedhomeip/issues/40781)

#### Testing
CI
 
#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

pcaddict committed:
Resolve ambiguous operator overload

resolves an ambiguous operator== comparison between a base class (ConcreteAttributePath) and a derived class (ConcreteDataAttributePath) by adding an explicit static_cast.

pcaddict committed:
Resolves compiler warning for compound assignment to volatile.

Refactors compound assignment to its explicit form.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=40794) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=40794) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=40794) it.</sub>

soares-sergio commented on src/app/ChunkedWriteCallback.cpp:
```diff
@@ -81,7 +81,7 @@ bool ChunkedWriteCallback::IsAppendingToLastItem(const ConcreteDataAttributePath
         return false;
     }
-    if (!mProcessingAttributePath.HasValue() || !(mProcessingAttributePath.Value() == aPath))
+    if (!mProcessingAttributePath.HasValue() || !(mProcessingAttributePath.Value() == static_cast<ConcreteAttributePath>(aPath)))
```
This seems like an easy mistake to make...

I wonder if we can add/overload something to the class itself to prevent making this mistake again?

pcaddict replied:
I've dug a little deeper and believe the solution can be improved, but I am not sure how widespread the ramifications are in terms of how this might apply elsewhere, if it matters.

A better solution may be to make the constructor for `ConcreteDataAttributePath` explicit. This prevents the implicit casts and removes the need for the explicit casts added in my PR. On the other hand, there are two other places that will then need explicit casts to resolve situations where implicit casts were used in function calls.

For now I will leave the new diff in this comment for discussion. I am unsure what the correct path forward is in pushing these changes, or even if this is the desired solution but I think it is worth consideration. 

```
diff --git a/src/app/ChunkedWriteCallback.cpp b/src/app/ChunkedWriteCallback.cpp
index 6ec1d07..bae45ba 100644
--- a/src/app/ChunkedWriteCallback.cpp
+++ b/src/app/ChunkedWriteCallback.cpp
@@ -40,7 +40,7 @@ void ChunkedWriteCallback::OnResponse(const WriteClient * apWriteClient, const C
         }
 
         // This is a response to another attribute write. Report the final result of last attribute write.
-        callback->OnResponse(apWriteClient, mProcessingAttributePath.Value(), mAttributeStatus);
+        callback->OnResponse(apWriteClient, static_cast<ConcreteDataAttributePath>(mProcessingAttributePath.Value()), mAttributeStatus);
     }
 
     // This is the first report for a new attribute.  We assume it will never be a list item operation.
@@ -66,7 +66,7 @@ void ChunkedWriteCallback::OnDone(WriteClient * apWriteClient)
         // We have a cached status that has yet to be reported to the application so report it now.
         // If we failed to receive the response, or we received a malformed response, OnResponse won't be called,
         // mProcessingAttributePath will be Missing() in this case.
-        callback->OnResponse(apWriteClient, mProcessingAttributePath.Value(), mAttributeStatus);
+        callback->OnResponse(apWriteClient, static_cast<ConcreteDataAttributePath>(mProcessingAttributePath.Value()), mAttributeStatus);
     }
 
     mProcessingAttributePath = NullOptional;
@@ -81,7 +81,7 @@ bool ChunkedWriteCallback::IsAppendingToLastItem(const ConcreteDataAttributePath
     {
         return false;
     }
-    if (!mProcessingAttributePath.HasValue() || !(mProcessingAttributePath.Value() == static_cast<ConcreteAttributePath>(aPath)))
+    if (!mProcessingAttributePath.HasValue() || !(mProcessingAttributePath.Value() == aPath))
     {
         return false;
     }
diff --git a/src/app/ConcreteAttributePath.h b/src/app/ConcreteAttributePath.h
index 99afb46..48bb1f9 100644
--- a/src/app/ConcreteAttributePath.h
+++ b/src/app/ConcreteAttributePath.h
@@ -113,7 +113,7 @@ struct ConcreteDataAttributePath : public ConcreteAttributePath
 
     ConcreteDataAttributePath() {}
 
-    ConcreteDataAttributePath(const ConcreteAttributePath & path) : ConcreteAttributePath(path) {}
+    explicit ConcreteDataAttributePath(const ConcreteAttributePath & path) : ConcreteAttributePath(path) {}
 
     ConcreteDataAttributePath(EndpointId aEndpointId, ClusterId aClusterId, AttributeId aAttributeId) :
         ConcreteAttributePath(aEndpointId, aClusterId, aAttributeId)
diff --git a/src/app/WriteHandler.cpp b/src/app/WriteHandler.cpp
index 7fa782c..8566d24 100644
--- a/src/app/WriteHandler.cpp
+++ b/src/app/WriteHandler.cpp
@@ -324,7 +324,7 @@ namespace {
 // To reduce the various use of previousProcessed.HasValue() && previousProcessed.Value() == nextAttribute to save code size.
 bool IsSameAttribute(const Optional<ConcreteAttributePath> & previousProcessed, const ConcreteDataAttributePath & nextAttribute)
 {
-    return previousProcessed.HasValue() && previousProcessed.Value() == static_cast<ConcreteAttributePath>(nextAttribute);
+    return previousProcessed.HasValue() && previousProcessed.Value() == nextAttribute;
 }
 
 bool ShouldReportListWriteEnd(const Optional<ConcreteAttributePath> & previousProcessed, bool previousProcessedAttributeIsList,
```

pcaddict committed:
Use explicit constructor to resolve ambiguity

* Marks ConcreteDataAttributePath constructor explicit, protecting against implicit conversions.

* Add explicit casts at call sites previously relying on implicit base-to-derived conversion.

pcaddict committed:
Merge branch 'project-chip:master' into compiler_compatibility

pcaddict committed:
Revert commits in preparation for alternative solution

Reverts 0af10f1 and e946a32 to allow for less intrusive solution to be
implemented.

pcaddict committed:
Resolve ambiguous comparison with operator overload

Resolves a build ambiguity that occurs with C++20 compilers when
comparing a ConcreteAttributePath and a ConcreteDataAttributePath.

This commit adds a `operator==` to the derived class that accepts the
base class. This provides a better-ranked overload candidate, resolving
the ambiguity. The implementation delegates to the base class
`operator==`, preserving the exact C++17 behavior.

pcaddict committed:
Fix formatting to satisfy CI

Reformatted function to conform to Restyler's rules. No functional
changes, only formatting to satisfy CI.

bzbarsky-apple reviewed (approved):
My apologies for the lag @pcaddict .  I have been having a having a hard time keeping track of Github notifications, unfortunately.
